[DRAFT/RFC] Documentation expansion (README.md, CONTRIBUTING.md, CONVENTIONS.md)

Vaerian opened the pull request:
## Documentation Expansion

### Why

Fabric API is a **mess**. Well maybe not _that_ big of a **mess**. Maybe more of a mess, not a **mess**. Either way, there's been many recommendations to standardize contributions made to Fabric API, and an equal number of attempts. However, somewhere along the way, in an attempt to organize Fabric API, we became disorganized in the amount of attempts we made to get organized.

### What

This draft PR/RFC/Whatever you want to call it, is meant to act as a centralized place to get our footing and understand the scope of standardization. I'm not naïve to the fact that this will likely take ages to develop a consensus on these standards, and that it'll take many a revision to get to a place where people are happy.

That being said, although it's kind of annoying to deal with through GitHub, I'd like to use [Vaerian/fabric](https://github.com/Vaerian/fabric/tree/documentation-expansion) as like a central repo for figuring this out. I've enabled issues on the repo so larger more general discussion can happen here, and the specifics can happen in issues on the forked repo. If they request it I'm going to give i509VCB, Player, and modmuss50 push access to the repo (actually it sounds like people with push access to FabricMC/fabric can just straight up make commits to this PR, so it's probably not necessary, but the offer remains), as ultimately they have authority over whether this PR will get in or not, all other issues will get pulled in based on democratic discussion (again, to the best of my ability). PRs into [Vaerian/fabric on documentation-expansion](https://github.com/Vaerian/fabric/tree/documentation-expansion) are highly encouraged.

Here's to hoping something good comes out of this 🤞🏻 

### Are You Insane?

You're probably asking yourself, why would anyone ever subject themselves to managing a "sub repository" developing standards for an open source project? That sounds horrible. Here's the deal, I have like problematic levels of OCD and so bikeshedding is like my passion. Little side tangent, I was tasked with figuring out a GH Action for managing releases and all of that for a mod, that turned into me writing a command line tool for creating fabric projects and managing them, that turned into me writing a command line tool for managing a changelog, and then that turned into me writing a rust library for parsing command line input. The point is I thrive in areas that are organized and have standards and such. I once had a dream where I created an organization that standardized other standards organizations. So for me it's actually really fun going through the repository issues and compiling it all together, and managing figuring all of this out. Maybe that makes me masochist I'm not sure.

Please ping me `@Vaerian` on fabricord, or basically any other fabric related server (I think I'm on most of them at this point), if you have anything you wanna discuss. I'm never against a good ping.

### The Meat And Potatoes

As topic areas are brought up I’ll add them to the list up here. This PR should only be for outlining the standards, and writing the contribution guidelines, not for actually implementing it in Fabric API. Presumably a separate PR will be made later to align Fabric API with the standards officially set in place by this PR. Additionally, all future style related PRs should both add to the CONVENTIONS.md file and modify all of Fabric API to meet the new style.

PR scope
- [x] Closes #328 
- [x] Resolves #417
- [x] Includes / Resolves #499 
- [x] Includes / Resolves #575 
- [x] Includes / Resolves #510 
- [x] Includes / Resolves #587 
- [x] Includes / Resolves #637 
- [ ] More in depth CONTRIBUTING.md guide
- [ ] Revival of additional standards that have fallen to the wayside
- [ ] Addition of some de facto standards already in Fabric API
- [ ] Figure out the common namespace in a separate RFC and then add that to CONVENTIONS.md

This PR should replace #417 and close #328 along with standardizing the other related issues listed below. After this is pulled in an admin should take a look at all of the related issues and assess whether or not they should be closed or not.

Related issues
- [#587] Mixins shadowing final fields should use `@Final`
- [#637] Lets Talk: Inconsistent names for classes and mixins.
- [#986] minor mixin improvements
- [#328] Add a proper README.md
- [#417] Improved README
- [#510] Inconsistent method names for mixin methods
- [#531] Refactor several old names
- [#575] [RFC] New module naming scheme
- [#499] [RFC] Experimental APIs

Commit:
Wrote a more detailed and helpful README

Additionally, added empty files for a CONTRIBUTING.md and a CONVENTIONS.md to get started.

Closes #328 and closes #417

Commit:
Added more detailed information about the project to the README.md

Added examples of the kinds of tools fabric api provides, and listed links out to other projects and resources

Commit:
Added links to the fabric website and wiki in the README.md

Commit:
"Guess what!"
"What?"
"'Fabric' is a proper noun. Learn how to capitalize you big dummy"

Commit:
Initial CONVENTIONS.md

Compiled all of the discussions from various style issues into one markdown file outlining some of the standards for Fabric API contributions.

haykam821 commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+Multi-word names can be split or combined depending on what makes sense from a hierarchical standpoint. Take for example a module named `fabric-lifecycle-events-v1`. It would have the root package `net.fabricmc.fabric.api.event.lifecycle.v1` to maintain the hierarchy of lifecycle being a type of event. A module named `fabric-networking-block-entity-v1` would have the root package `net.fabricmc.fa ...
+
+Experimental apis in Fabric API ought to be annotated using `@Deprecated` in order to indicate instability and the potential for the api to drastically change between versions.
```
```suggestion
Experimental APIs in Fabric API ought to be annotated using `@Deprecated` in order to indicate instability and the potential for the API to drastically change between versions.
```

Vaerian replied:
https://github.com/FabricMC/fabric/pull/1062/commits/64385bd047e9c3653b36046edfdab23b5378ea7c

haykam821 commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+Regular mixin classes must be named the same name as the target class they are being mixed into, with "Mixin" appended to the end of the class name. As an example, a mixin class into the class `MinecraftServer` should be named `MinecraftServerMixin`.
+
+Accessor mixin classes must be named the same name as the target class they are being mixed into, with "Accessor" appended to the end of the class name. As an example, the accessor class into the class `MinecraftServer` should be named `MinecraftServerAccessor`.
```
```suggestion
Likewise, accessor mixin classes must be named the same name as the target class they are being mixed into, with "Accessor" appended to the end of the class name. As an example, the accessor class into the class `MinecraftServer` should be named `MinecraftServerAccessor`.
```

Vaerian replied:
Should be resolved in https://github.com/FabricMC/fabric/pull/1062/commits/64385bd047e9c3653b36046edfdab23b5378ea7c

Commit:
Added the start to CONTRIBUTING.md

The file is definitely not finished, I'll work on it later, just wanted a start up in case someone else wanted to take a stab at it.

i509VCB commented on README.md:
```diff
@@ -1,7 +1,36 @@
+
+For more detailed instructions, look [here](https://fabricmc.net/wiki/install).
+
+In addition, most mods depend on an API mod, containing commonly used hooks. It can be found [here](https://minecraft.curseforge.com/projects/fabric/files).
```
I'd probably also mention github releases here

i509VCB replied:
```suggestion
In addition, most mods depend on Fabric API. Fabric API can be downloaded [here](https://minecraft.curseforge.com/projects/fabric/files).
```

i509VCB commented on README.md:
```diff
@@ -1,7 +1,36 @@
 
 Essential hooks and patches for modding with Fabric.
 
+Fabric API is the library for essential hooks and interop mechanisms for Fabric mods. Examples include:
```
```suggestion
Fabric API is the library for essential hooks and interoperability mechanisms for Fabric mods. Examples include:
```

Full word here is better imo

i509VCB commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+## General
+
+All code should meet the Yarn naming conventions which can be found [here](https://github.com/FabricMC/yarn/blob/20w21a/CONVENTIONS.md).
```
Could you link this to 1.16.2 conventions for yarn.

i509VCB commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+Fabric API tends to be restricted to small, highly used, and breaking changes to the vanilla game. Examples include helpers for adding blocks, which are used by basically every modder ever (hence their inclusion); breaking changes to the rendering system, which ensure that mods can remain compatible with one another; or commonly used mixins that, because of their nature, would otherwise cause mod ...
+
+If your contribution is a massive system, or api for something that doesn't exist in the vanilla game, it does not belong in Fabric API. The best examples are things like, custom world generators, mana systems, or a fluid api. All of these things represent content expansions that don't exist in the base game and should not be included in Fabric API. That doesn't mean however, that your idea is no ...
```
The scope of whether a large subsystem belongs in Fabric API is definitely up to the discretion of API maintainers. Though the general rule there does apply. Further discussion on this is probably needed.

i509VCB replied:
FabLabs is primarily for incubating and more experimental concepts, such as screen handler migration or game rule api. These are directly applicable to vanilla itself but these require more work to finish.

i509VCB commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+## Style Check
+
+Talk about the checkstyle and the license formatter
```
Player wrote up something about code style in pretty succinct language. I'd probably mention the `licenseFormat` task somewhere for people who don't know about it.

i509VCB commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+## Taking The First Step
+
+The first step to contributing to Fabric API is making an issue. Head over to the [Fabric API GitHub repository](https://github.com/FabricMC/fabric/issues/) and create a detailed issue outlining what you'd like added to Fabric API. From there, the community can weigh in on whether or not that idea is within the scope of Fabric API, how to best implement the idea, and any additional details that n ...
```
I wouldn't call it required, but a little closer to "You are encouraged to make an issue".

i509VCB commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+The first step to contributing to Fabric API is coming up with an idea, and while all ideas may be useful, not all ideas necessarily fit within the scope of Fabric API.
+
+Fabric API tends to be restricted to small, highly used, and breaking changes to the vanilla game. Examples include helpers for adding blocks, which are used by basically every modder ever (hence their inclusion); breaking changes to the rendering system, which ensure that mods can remain compatible with one another; or commonly used mixins that, because of their nature, would otherwise cause mod ...
```
Sentence here is really long and should probably be split up a bit.

I'd use `mod developer` over `modder` for a more professional sounding tone. I'd probably mention that features which are hard to implement correctly (such as networking api) are within scope.

i509VCB commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+Before we begin, some friendly advice:
+1. We don't tend to ban people just for having ideas, no matter how outlandish. However, “no” can mean “no”, and un-constructive responses to constructive arguments can get on our nerves after some time. We are generally unpaid volunteers who dedicate a lot of free time to the tiring job of social interaction - please don't make our work harder than it already is.
+2. We're here to help each other learn, after all. Don't hesitate to talk to us at any point! [The Fabric Discord](https://discord.gg/v6v4pMv) is a great place to ask questions and get input on your contributions.
+3. Fabric isn't easy to get your content into, as we uphold standards of quality and are very paranoid about making a bad precedent. You will be scrutinized again, and again, and again; you will be asked questions about more-or-less every line in your pull request (even the imports!).
```
Second sentence here probably isn't necessary. Paranoia probably isn't the intent either.

I would probably word this along the line of, "Contributions to Fabric API are expected to hold a certain standard of quality. These standards are justified in order to provide a reasonably stable API for mods to build off of. You should be prepared to answer questions related to features you propose and the implementation detail of said features."

i509VCB commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+## The Hard Part
+
+Idk write some stuff about how to actually write code and test it in the fapi environment
```
I would probably link to code conventions here.

i509VCB commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+Talk about the review process
+
+## Last Call
```
For future write up, it should be mentioned that bug fixes, unless breaking api do not apply to Last call

i509VCB commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+## Annotations
+
+All methods that can potentially return a null value should be annotated in-line using the JetBrains `@Nullable` right before the return type definition for the method. For example, a method named `possiblyGetName` that can potentially return a null value should be written as `public @Nullable String possiblyGetName()...`.
```
Arguably this is a whole entire issue in itself here. Unsure if this should be handled in a future pull request

i509VCB commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+## Enums
+
+Enum values that are referenced internally (values referenced within methods associated with their own enum definition) must always use `Enum.VALUE` never simply `VALUE`.
```
I am arguably in favor of requiring that you qualify references to enum members, but this is should probably be part of a future pull request for actual conventions.

i509VCB commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+Multi-word names can be split or combined depending on what makes sense from a hierarchical standpoint. Take for example a module named `fabric-lifecycle-events-v1`. It would have the root package `net.fabricmc.fabric.api.event.lifecycle.v1` to maintain the hierarchy of lifecycle being a type of event. A module named `fabric-networking-block-entity-v1` would have the root package `net.fabricmc.fa ...
+
+Experimental APIs in Fabric API ought to be annotated using `@Deprecated` in order to indicate instability and the potential for the API to drastically change between versions.
```
I would probably mention a bit of context about Experimental APIs before mentioning this.

i509VCB commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+Every name consisting of multiple words must have a hypen (-) between the words. This includes class names such as BlockEntity being represented as `block-entity`.
+
+Packages must follow the following format (using `fabric-biome-api-v1` as an example). API Packages must use the root `net.fabricmc.fabric.api.biome.v1`. Implementation Packages must use the root `net.fabricmc.fabric.impl.biome`. Mixin Packages must use the root `net.fabricmc.fabric.mixin.biome`.
```
Client packages should be mentioned here as their standard is slightly different.

i509VCB commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+Mixin injection and redirect methods must be named based upon the specific function of the injection or redirection, not the method that they are injecting into or redirecting. For example a method inject that registers commands on the dedicated server ought to be named `registerCommands` not the name of the target method, in this case `setupServer`.
+
+If a shadowed method needs to be prefixed because of a conflict it must use the prefix `minecraft$` for methods in minecraft, `[namespace]$` for projects with a different namespace (which should always match the mod id), or if neither of those are available `[package name]$`. Take for example a situation where the shadowed method `fromJson` conflicts with a new `fromJson` method. The shadowed met ...
```
This would occur very rarely, but I would lean towards the default prefix of `shadow$`.

That explicitly specifies that the method being shadowed in name. Sponge guys have had pretty good milage with that default prefix for everything, but fabric api would probably use it for conflicts only.

i509VCB commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,55 @@
+
+If a shadowed method needs to be prefixed because of a conflict it must use the prefix `minecraft$` for methods in minecraft, `[namespace]$` for projects with a different namespace (which should always match the mod id), or if neither of those are available `[package name]$`. Take for example a situation where the shadowed method `fromJson` conflicts with a new `fromJson` method. The shadowed met ...
+
+All mixin classes soft implementing interfaces must use the prefix `fabric$` and the `@Implements` annotation from the mixin library. For example, take a class that soft implements the interface `Identifiable`, it must be annotated as follows `@Implements(@Interface(iface = Identifiable.class, prefix = "fabric$"))`.
```
`fabric$` specifier is probably not right for every context. This specific case probably belongs in a future pull request.

i509VCB commented on README.md:
```diff
@@ -1,7 +1,36 @@
-To setup a Fabric development environment, check out the [example fabric mod](https://github.com/FabricMC/fabric-example-mod) and follow the instructions there.
+### For Players
+
+To install Fabric, you need the **Fabric Loader**, which you can download [here](https://fabricmc.net/use).
```
```suggestion
To install Fabric, you need to have **Fabric Loader** installed. Downloading and instructions for installing Fabric Loader can be found [here](https://fabricmc.net/use).
```

i509VCB commented on README.md:
```diff
@@ -1,7 +1,36 @@
+
+To setup a Fabric development environment, check out the [example Fabric mod](https://github.com/FabricMC/fabric-example-mod) and follow the instructions there.
+
+The Fabric wiki also has a series of helpful tutorials to teach you how to create your first mod. They can be found [here](https://fabricmc.net/wiki/start#tutorials2).
```
```suggestion
The [Fabric wiki](https://fabricmc.net/wiki/start#tutorials2) has a series of helpful tutorials to teach you the basics of mod development. The wiki includes tutorials on topics such as registering blocks, items and more advanced topics such as commands.
```

i509VCB commented on README.md:
```diff
@@ -1,7 +1,36 @@
+
+### For Contributors
+
+If you're looking to contribute to Fabric API take a look at the [contributions page](CONTRIBUTING.md) for information on how to get started.
```
There is probably a more inviting way to word this.

i509VCB reviewed (changes requested):
I do respect the fact you put in what looks to be plenty of time into writing this. However I feel we may need to decrease scope with this pull request since this covers so many topics. Since we somewhat working with prior art here, I feel there are parts of what I've written and Player has written that could be valuable here.

One thing I would do is reread the entire document and think of the following as I read:
- Does this make any sense to someone with no knowledge of how contribution process works
here.
- Is the wording I am using to describe these concepts clearly. Especially important considering many people don't speak English as a first language or machine translate these things to figure out things.
- Do I describe the processes in depth while being succinct and without indirection.

ghost commented:
Yeah I'd be more than happy to take what you and Player have put together and fit that into something. I agree the scope is kind of large, but I feel like it's kind of snowballed into that, and the easiest way right now to address it would be to do it in bulk.

haykam821 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+Fabric API tends to be restricted to small, highly used, and breaking changes to the vanilla game. Examples include helpers for adding blocks, which are used by basically every modder ever (hence their inclusion); breaking changes to the rendering system, which ensure that mods can remain compatible with one another; or commonly used mixins that, because of their nature, would otherwise cause mod ...
+
+If your contribution is a massive system, or api for something that doesn't exist in the vanilla game, it does not belong in Fabric API. The best examples are things like, custom world generators, mana systems, or a fluid api. All of these things represent content expansions that don't exist in the base game and should not be included in Fabric API. That doesn't mean however, that your idea is no ...
```
```suggestion
If your contribution is a massive system, or API for something that doesn't exist in the vanilla game, it does not belong in Fabric API. The best examples are things like, custom world generators, mana systems, or a fluid api. All of these things represent content expansions that don't exist in the base game and should not be included in Fabric API. That doesn't mean however, that your idea is not useful. Larger, more experimental libraries are being developed over at [FabLabs](https://github.com/FabLabsMC). They might be interested in your idea.
```

Scotsguy replied:
```suggestion
If your contribution is a massive system or API for something that doesn't exist in the vanilla game, it does not belong in Fabric API. The best examples are things like custom world generators, mana systems, or a fluid API. All of these things represent content expansions that don't exist in the base game and should not be included in Fabric API. However, that doesn't mean that your idea is not useful. Larger, more experimental libraries are being developed over at [FabLabs](https://github.com/FabLabsMC). They might be interested in your idea.
```
